arduino-core/src/cc/arduino/packages/uploaders/SerialUploader.java

```diff
@@ -393,13 +374,27 @@ public boolean burnBootloader() throws Exception {
 
     new LoadVIDPIDSpecificPreferences().load(prefs);
 
-    String pattern = prefs.getOrExcept("erase.pattern");
-    String[] cmd = StringReplacer.formatAndSplit(pattern, prefs);
-    if (!executeUploadCommand(cmd))
+    if (!runCommand("erase.pattern", prefs))
       return false;
 
-    pattern = prefs.getOrExcept("bootloader.pattern");
-    cmd = StringReplacer.formatAndSplit(pattern, prefs);
-    return executeUploadCommand(cmd);
+    return runCommand("bootloader.pattern", prefs);
+  }
+
+  private boolean runCommand(String patternKey, PreferencesMap prefs) throws Exception, RunnerException {
+    try {
+      String pattern = prefs.getOrExcept(patternKey);
+      StringReplacer.checkIfRequiredKeyIsMissingOrExcept("serial.port", pattern, prefs);
+      String[] cmd = StringReplacer.formatAndSplit(pattern, prefs);
+      return executeUploadCommand(cmd);
+    } catch (RunnerException e) {
+      throw e;
+    } catch (PreferencesMapException e) {
+      if (e.getMessage().equals("serial.port")) {
+        throw new SerialNotFoundException(e);
```

Comment:
This seems like a slightly misleading exception name. AFAICS the serial port is not *set*, SerialNotFound suggests (to me) that the configured port is not found on the system.

Reply:
Ah, the exception is used for both "no port selected" and "selected port not found". I guess that makes sense, especially if the handling is the same (prompt the user for a port), so splitting them probably does not make much sense (if we need to differentiate between these cases, the exception could always get a "port" attribute that stores the failing port, so a catcher can check the attribute for null to see that no port was selected at all).